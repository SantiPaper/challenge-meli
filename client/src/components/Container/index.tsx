import { FC, PropsWithChildren } from "react";
import { StyledContainer } from "./style";

export const Container: FC<PropsWithChildren> = ({ children }) => {
  return <StyledContainer>{children}</StyledContainer>;
};
