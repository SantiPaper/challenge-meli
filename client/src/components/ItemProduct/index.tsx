import { StyledItemProduct } from "./style";
import imgShipping from "../../../public/images/shipping.png";

type Props = {
  title: string;
  price: number;
  img: string;
  shipping: boolean;
  id: string;
};
export const ItemProduct = ({ title, price, img, shipping, id }: Props) => {
  return (
    <StyledItemProduct id={id}>
      <img
        className="product__img"
        src={img}
        alt={title}
        width={180}
        height={180}
      />

      <div className="product__info">
        <div className="product__info__price-container">
          <p className="product__info__price-container__price">$ {price}</p>
          {shipping && <img src={imgShipping} alt="" width={15} height={15} />}
        </div>
        <h3 className="product__info__title">{title}</h3>
      </div>
    </StyledItemProduct>
  );
};
