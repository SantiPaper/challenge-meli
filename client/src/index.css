@font-face {
  font-family: "Proxima nova";
  src: url("./fonts/proximanova_regular.ttf");
  font-weight: 400;
}

@font-face {
  font-family: "Proxima nova";
  src: url("./fonts/proximanova_bold.otf");
  font-weight: 600;
}

* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

html {
  font-size: 10px;
  font-family: "Proxima nova";
  background-color: var(--gray-100);
}

:root {
  --yellow: #ffe600;
  --gray-900: #333333;
  --gray-600: #666666;
  --gray-400: #999999;
  --gray-100: #eeeeee;
  --blue: #3483fa;

  --spacing-m: 1.6rem;
  --spacing-l: 3.2rem;
  --spacing-xl: 5rem;

  --font-size-100: 1.2rem;
  --font-size-200: 1.4rem;
  --font-size-300: 1.6rem;
  --font-size-400: 1.8rem;
  --font-size-500: 2.4rem;
  --font-size-600: 2.8rem;
  --font-size-700: 4.6rem;

  --radius: 0.4rem;
}

.sr-only {
  clip: rect(0 0 0 0);
  border: 0;
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}
