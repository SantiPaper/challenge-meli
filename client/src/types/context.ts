import { SearchResults } from "./items";

export type Context = {
  state: State;
  getProductsBySearch: (searchResults: SearchResults) => void;
  getSearchString: (query: string) => void;
};

export type State = {
  searchResults: SearchResults;
  searchInput: string;
};
export type Reducer = (state: State, action: Action) => State;

export type Action =
  | {
      type: "GET_PRODUCTS";
      payload: SearchResults;
    }
  | {
      type: "SEARCH_PRODUCTS";
      payload: string;
    };
