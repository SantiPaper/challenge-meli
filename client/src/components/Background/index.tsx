import { FC, PropsWithChildren } from "react";
import { Container } from "../Container";
import { StyledBackground } from "./style";

export const Background: FC<PropsWithChildren> = ({ children }) => {
  return (
    <Container>
      <StyledBackground>{children}</StyledBackground>
    </Container>
  );
};
