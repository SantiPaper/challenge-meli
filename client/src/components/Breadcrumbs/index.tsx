import { Categories } from "../../types/items";
import { StyledBreadcrumbs } from "./style";

type Props = {
  categories: Categories;
};
export const Breadcrumbs = ({ categories }: Props) => {
  return (
    <StyledBreadcrumbs>
      <ul className="breadcrumbs">
        {categories.map((catergory) => (
          <li key={catergory} className="breadcrumbs__item">
            <a href="#" className="breadcrumbs__item__link">
              {catergory}
            </a>
          </li>
        ))}
      </ul>
    </StyledBreadcrumbs>
  );
};
