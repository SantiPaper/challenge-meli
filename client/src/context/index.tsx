import { PropsWithChildren, createContext, useReducer } from "react";
import { Context } from "../types/context";
import { initialState, reducer } from "./reducer";
import { ProductDetail, SearchResults } from "../types/items";

export const ItemsContext = createContext<Context | null>(null);

export const Provider = ({ children }: PropsWithChildren) => {
  const handleFormat = (price: number) => {
    return new Intl.NumberFormat("es-AR", {
      style: "currency",
      currency: "ARS",
      minimumFractionDigits: 0,
      maximumFractionDigits: 0,
    }).format(price);
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const getProductsBySearch = (searchResults: SearchResults) => {
    dispatch({
      type: "GET_PRODUCTS",
      payload: searchResults,
    });
  };

  const getProductById = (product: ProductDetail) => {
    dispatch({
      type: "GET_PRODUCT_BY_ID",
      payload: product,
    });
  };

  return (
    <ItemsContext.Provider
      value={{
        state,
        getProductsBySearch,
        getProductById,
        handleFormat,
      }}
    >
      {children}
    </ItemsContext.Provider>
  );
};
