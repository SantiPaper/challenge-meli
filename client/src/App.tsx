import { Routes, Route } from "react-router-dom";
import { Home } from "./components/Home";
import { Results } from "./components/Results";
import { Detail } from "./components/Detail";
import { Layout } from "./components/Layout";
function App() {
  return (
    <Layout>
      <Routes>
        <Route path="/" element={<Home />} />
        <Route path="/items" element={<Results />} />
        <Route path="/items/:id" element={<Detail />} />
      </Routes>
    </Layout>
  );
}

export default App;
