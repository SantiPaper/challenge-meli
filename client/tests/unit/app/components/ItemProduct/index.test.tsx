import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom"; // Importa BrowserRouter desde react-router-dom
import { mockedProduct } from "./__mocks__";
import { ItemProduct } from "../../../../../src/components/ItemProduct";
import { Provider } from "../../../../../src/context";

beforeEach(() => {
  render(
    <BrowserRouter>
      <Provider>
        <ItemProduct
          condition={mockedProduct.condition}
          id={mockedProduct.id}
          img={mockedProduct.img}
          price={mockedProduct.price}
          shipping={mockedProduct.shipping}
          title={mockedProduct.title}
          key={mockedProduct.id}
        />
      </Provider>
    </BrowserRouter>
  );
});

describe("ItemProduct", () => {
  test("Renderiza bien", () => {
    const { container } = render(
      <BrowserRouter>
        <Provider>
          <ItemProduct
            condition={mockedProduct.condition}
            id={mockedProduct.id}
            img={mockedProduct.img}
            price={mockedProduct.price}
            shipping={mockedProduct.shipping}
            title={mockedProduct.title}
            key={mockedProduct.id}
          />
        </Provider>
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });
  test("Verifica que el enlace apunte a la ruta correcta", () => {
    const linkElement = screen.getByRole("link");
    expect(linkElement).toHaveAttribute("href", `/items/${mockedProduct.id}`);
  });
  test("Se muestra el precio", () => {
    const price = screen.getByText(mockedProduct.price);
    expect(price).toBeInTheDocument();
  });
  test("Se muestra el titulo", () => {
    const title = screen.getByText(mockedProduct.title);
    expect(title).toBeInTheDocument();
  });
  test("Se muestra la condiciÃ³n", () => {
    const condition = screen.getByTestId("condition-text");
    expect(condition).toBeInTheDocument();
  });
});
