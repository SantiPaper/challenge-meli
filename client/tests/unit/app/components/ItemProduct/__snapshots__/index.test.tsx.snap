// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ItemProduct > Renderiza bien 1`] = `
<div>
  <article
    class="sc-beySPh duuIyI"
    data-testid="Producto de prueba"
  >
    <a
      href="/items/1"
    >
      <img
        alt="Producto de prueba"
        class="product__img"
        height="180"
        src="imagen.jpg"
        width="180"
      />
      <div
        class="product__info"
      >
        <div>
          <div
            class="product__info__price-container"
          >
            <p
              class="product__info__price-container__price"
            >
              $100
            </p>
            <img
              alt=""
              height="15"
              src="/public/images/shipping.png"
              width="15"
            />
          </div>
          <div>
            <h3
              class="product__info__title"
            >
              Producto de prueba
            </h3>
          </div>
        </div>
        <p
          class="product__info__condition"
          data-testid="condition-text"
        >
          Nuevo
        </p>
      </div>
    </a>
  </article>
</div>
`;
